### DISCUSSION TITLE
DNS not resolving in iPXE script after manual interface config

### DISCUSSION THREAD

**Discussion by isaacdavis** [ipxe/ipxe#1321] [Q&A] [answered]
Hi there,

We at Fastly are using iPXE to boot servers, with some constraints:
* We cannot assign an IP address using DHCP -- DHCP is available for testing but not in production.
  * We are instead generating a script in advance with the mac address of the desired interface, and the desired IP address, and iterating through the available interfaces until a match is found, at which point we manually configure the interface and proceed with the boot.
* We have to use a public (as opposed to local) DNS server -- a local DNS server is available for testing but not in production.
* Boot images are served with mTLS.

We've run into an interesting issue where DNS won't resolve from the script, but if I drop to a shell from the script, DNS starts working. Here's the script I'm working with -- I've had to make some redactions, but I hope the intent is clear:

```
#!ipxe

# mac address of desired interface (example)
set boot_mac:hex 01:23:45:67:89:ab

# loop through ifaces until a matching mac is found
set curr_iface_idx:int32 0
:iface_loop isset ${net${curr_iface_idx}/mac} || goto no_match_found

    set boot_iface net${curr_iface_idx}
    iseq ${${boot_iface}/mac} ${boot_mac} && goto match_found || echo ${boot_iface} does not match ${boot_mac}
    inc curr_iface_idx && goto iface_loop

:no_match_found
echo no iface found with mac ${boot_mac}. Cannot configure interfaces.
prompt --timeout 10000 Press any key for iPXE command line... && shell || exit

# Matching iface found. Manually configure iface.
:match_found
echo ${boot_iface} matches ${boot_mac}
# Example IP and gateway
set ${boot_iface}/ip 10.253.140.9
set ${boot_iface}/netmask 255.255.0.0
set ${boot_iface}/gateway 10.253.140.8

ifopen ${boot_iface}
iflinkwait ${boot_iface} --timeout 60 || echo link on ${boot_iface} timed out
set dns 8.8.8.8

# Ping the (redacted) URL to test DNS.
# This ping times out -- but in the shell that follows, an identical ping works just fine!
echo pinging once
ping --count 1 ipxe.example.com || shell

# Same here -- if I remove the above ping, this call to `initrd` times out, with repeated ellipses, not a percentage counting up.
# It works just fine in the shell that follows.
initrd https://ipxe.example.com/initrd.img || shell
chain https://ipxe.example.com/kernel initrd=initrd.img || shell
```

So, in summary, whatever command comes after the `set dns` command, whether a `ping` or an `initrd`, times out, but works fine from the shell that is spawned immediately after.

I've tried a few things:
* Different DNS servers (local and public)
* pinging multiple times
* sleeping after configuring dns
. . . and none of the above work.

I _have_ found that configuring with `ifconf` makes the script work fine. `ifconf` sets the DNS server to `10.253.0.1`, which does not work if I manually configure it. So, we know that DNS _can_ work, and that the calls to `initrd` and `chain` can work as well. Is `ifconf` doing something I'm not?

Any and all help would be much appreciated. Thank you, all!

**Comment by NiKiZe** [comment]
Any chance STP or similar is enabled on the ports?
Making it block for X seconds after link before allowing traffic to flow?

**Reply by mcb30** [reply, marked as answer]
It does sound plausibly like an STP, 802.1x, or similar issue.  For some of these mechanisms, the switch port timer will be initiated by the first broadcast packet, which would explain why sleeping before sending the ping has no effect.

You can try `DEBUG=dns:3,udp:3,netdevice:3` to see some hopefully relevant debug messages for every packet that iPXE sends or receives.

**Reply by isaacdavis** [reply]
Thank you both -- you both were correct!

* I verified that disabling STP made the ping work
* I changed the script to ping once to initiate the port timer, sleep 10 seconds, then ping again. The first ping failed and the second ping went through, after which I was able to pull the kernel and boot.

Marking this as solved. Thank you both, again.

**Reply by mcb30** [reply]
Good to know, thanks for reporting back. 

We do try to handle STP and similar protocols automatically and transparently as far as possible, but that relies on having some kind of transaction that starts with a network broadcast and allows for retries (e.g. DHCP).  With a static IP configuration such as you are using, there's no opportunity for iPXE to defer completion, so the workaround you have implemented is probably the best available.
